modules/ingester/flush.go

```diff
@@ -52,16 +61,60 @@ func (i *Ingester) Flush() {
 	}
 }
 
-// FlushHandler calls sweepUsers(true) which will force push all traces into the WAL and force
+// ShutdownHandler handles a graceful shutdown for an ingester. It does the following things in order
+// * Stop incoming writes by exiting from the ring
+// * Flush all blocks to backend
+func (i *Ingester) ShutdownHandler(w http.ResponseWriter, _ *http.Request) {
+	// stop accepting new writes
+	err := i.markUnavailable()
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		_, _ = w.Write([]byte("error marking ingester unavailable"))
+	}
+
+	// move all data into flushQueue
+	i.sweepAllInstances(true)
+
+	// wait for blocks to complete and flush
+	for i.isWaitingForFlush() {
+		time.Sleep(100 * time.Millisecond)
+	}
```

Comment:
elsewhere you are using chans/atomic ints to track state.  i believe it's so that you won't move past this part of the shutdown handler until everything is flushed.  we need to find a simpler way to do this by inspecting the state of the queue itself.  these state variables are possibly going to expose us to some weird and difficult to debug race conditions.

something like:
```
while <len of queue> > 0 
  sleep a bit
  sweepAllInstances(true)   // put more stuff in the queue? 
```

we need to find a way to examine existing state to determine if we should move on if at all possible.